pkg/agent/docker_secrets_postrenderer.go

```diff
@@ -54,3 +76,105 @@ func (r *DockerSecretsPostRenderer) Run(renderedManifests *bytes.Buffer) (modifi
 
 	return modifiedManifests, nil
 }
+
+// updatePodSpecWithPullSecrets updates the podSpec inline with the relevant pull secrets.
+// We do not parse the yaml into actual Kubernetes objects since we want to be
+// independent of api versions. This requires special care and limitations, so
+// we limit our assumptions of the untyped handling to the following, with
+// clear errors when the assumption fails:
+// - The pod spec includes a 'containers' key with a slice value (with a clear error if this is not the case)
+// - Each container value is a map with an 'image' key and string value.
+func (r *DockerSecretsPostRenderer) updatePodSpecWithPullSecrets(podSpec map[interface{}]interface{}) error {
+	containersObject, ok := podSpec["containers"]
+	if !ok {
+		return fmt.Errorf("pod spec did not include a containers key: %+v", podSpec)
+	}
+	containers, ok := containersObject.([]interface{})
+	if !ok {
+		return fmt.Errorf("pod spec did not include a slice for the containers value: %+v", podSpec)
+	}
+
+	// If there are existing pull secrets, initialise our slice with that value
+	// and additionally initialize a map keyed by secret name which we can
+	// use to test existence more easily.
+	var imagePullSecrets []map[string]interface{}
+	existingNames := map[string]bool{}
+	if existingPullSecrets, ok := podSpec["imagePullSecrets"]; ok {
+		imagePullSecrets = existingPullSecrets.([]map[string]interface{})
+		for _, s := range imagePullSecrets {
+			if name, ok := s["name"]; ok {
+				if n, ok := name.(string); ok {
+					existingNames[n] = true
+				}
+			}
+		}
+	}
+
+	for _, c := range containers {
+		container, ok := c.(map[interface{}]interface{})
+		if !ok {
+			return fmt.Errorf("pod containers did not contain maps: %+v", c)
+		}
+		image, ok := container["image"].(string)
+		if !ok {
+			// NOTE: in https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#container-v1-core
+			// the image is optional to allow higher level config management to default or override (such as
+			// deployments or statefulsets), but both only define pod templates which in turn define containers?
+			return fmt.Errorf("pod container did not contain an image reference: %+v", container)
+		}
+
+		// Ignore (dockerhub) image refs without a domain.
+		domainSplit := strings.Split(image, "/")
+		if len(domainSplit) == 1 {
+			continue
+		}
+		imageDomain := domainSplit[0]
+
+		secretName, ok := r.secrets[imageDomain]
```

Comment:
I am not sure if we are just interested in a domain? For example, for a gcr container (`gcr.io/my_project/image:tag`), should we check `gcr.io` or `gcr.io/my-project`?

Reply:
Docker credentials are per server (though it turns out this can be the domain name or the fqdn of the server).